File: android/drengr/src/test/kotlin/dev/drengr/sdk/IdentifyTest.kt
package dev.drengr.sdk

import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.json.JSONObject
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File
import java.nio.file.Files
import java.util.concurrent.TimeUnit

/** Exercises identify()/setExperiment() against a real IngestSink flush over a
 *  MockWebServer (same client — OkHttp — the sink actually uses in production). */
class IdentifyTest {
    private val server = MockWebServer()

    private fun newSink(): IngestSink {
        server.start()
        return IngestSink(
            Files.createTempDirectory("drengr-test").toFile(),
            server.url("/ingest").toString(),
            "drengr_pk_test",
            mapOf("app_package" to "test", "install_id" to "test"),
            maxBatch = 1, // flush on the first enqueued event
        )
    }

    @After
    fun tearDown() {
        server.shutdown()
    }

    private fun awaitEnvelope(): JSONObject {
        val req = server.takeRequest(5, TimeUnit.SECONDS)
        assertNotNull("server received a request", req)
        server.enqueue(MockResponse().setResponseCode(200))
        return JSONObject(req!!.body.readUtf8())
    }

    private fun sampleEvent() = NetworkEvent(
        method = "GET", url = "https://api.example.com/x", statusCode = 200,
        durationMs = 5L, requestBodyBytes = 0L, responseBodyBytes = 2L,
        requestHeaders = emptyMap(), responseHeaders = emptyMap(),
        requestBody = null, responseBody = null, errorText = null,
        timestampMs = System.currentTimeMillis(),
    )

    @Test
    fun identifyEmitsEventAndAttachesExternalId() {
        server.enqueue(MockResponse().setResponseCode(200))
        val sink = newSink()
        sink.identify("user_123", mapOf("email" to "a@b.com", "plan" to "pro"))
        val envelope = awaitEnvelope()
        assertEquals("user_123", envelope.getString("external_id"))

        val events = envelope.getJSONArray("events")
        var identifyEv: JSONObject? = null
        for (i in 0 until events.length()) {
            val e = events.getJSONObject(i)
            if (e.getString("kind") == "identify") identifyEv = e
        }
        assertNotNull("identify event present", identifyEv)
        assertEquals("user_123", identifyEv!!.getString("external_id"))
        assertFalse("raw email must not appear", identifyEv.toString().contains("a@b.com"))
        assertTrue("non-PII trait survives", identifyEv.toString().contains("pro"))
    }

    @Test
    fun emptyExternalIdIsNoOp() {
        server.enqueue(MockResponse().setResponseCode(200))
        val sink = newSink()
        sink.identify("", mapOf("email" to "a@b.com"))
        assertEquals(null, server.takeRequest(300, TimeUnit.MILLISECONDS))
    }

    @Test
    fun externalIdAttachesToASubsequentEvent() {
        server.enqueue(MockResponse().setResponseCode(200))
        val sink = newSink()
        sink.identify("user_456")
        awaitEnvelope() // the identify event's own flush
        sink.addNetwork(sampleEvent())
        val envelope = awaitEnvelope()
        assertEquals("user_456", envelope.getString("external_id"))
    }

    @Test
    fun setExperimentAttachesAndClears() {
        server.enqueue(MockResponse().setResponseCode(200))
        val sink = newSink()
        sink.setExperiment("checkout_flow", "variant_b")
        sink.addNetwork(sampleEvent())
        var envelope = awaitEnvelope()
        assertEquals("variant_b", envelope.getJSONObject("experiments").getString("checkout_flow"))

        sink.setExperiment("checkout_flow", null)
        sink.addNetwork(sampleEvent())
        envelope = awaitEnvelope()
        assertFalse("experiments omitted once empty", envelope.has("experiments"))
    }

    @Test
    fun setExperimentBadInputIsNoOp() {
        server.enqueue(MockResponse().setResponseCode(200))
        val sink = newSink()
        sink.setExperiment("", "x") // must not throw
        sink.addNetwork(sampleEvent())
        val envelope = awaitEnvelope()
        assertFalse(envelope.has("experiments"))
    }
}

File: flutter/test/identify_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:drengr_flutter_sdk/drengr_flutter_sdk.dart';
import 'package:test/test.dart';

/// identify()/setExperiment() live on [IngestSink] (not [Drengr]) — this SDK's
/// capture engine and delivery sink are wired up by the app, not coupled here
/// (see tool/stall_probe.dart for the real wiring pattern).
void main() {
  group('identify + experiments', () {
    late HttpServer server;
    final bodies = <Map<String, dynamic>>[];

    setUp(() async {
      // Kill any queue persisted by earlier tests/runs (parallel isolates share it).
      final q = File(
          '${Directory.systemTemp.path}${Platform.pathSeparator}.drengr_queue.jsonl');
      if (q.existsSync()) q.deleteSync();
      bodies.clear();
      server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
      server.listen((req) async {
        final raw = await utf8.decoder.bind(req).join();
        bodies.add(jsonDecode(raw) as Map<String, dynamic>);
        req.response.statusCode = 200;
        await req.response.close();
      });
    });

    tearDown(() async {
      await server.close(force: true);
    });

    IngestSink newSink() => IngestSink(
          url: 'http://127.0.0.1:${server.port}/ingest',
          publishableKey: 'drengr_pk_test',
          context: {'app_package': 'test', 'install_id': 'test'},
          maxBatch: 1, // flush on the first enqueued event
          queuePath: '${Directory.systemTemp.path}${Platform.pathSeparator}'
              '.drengr_q_identify_${server.port}.jsonl', // isolate from parallel suites
        );

    NetworkEvent sampleEvent() => const NetworkEvent(
          method: 'GET',
          url: 'https://api.example.com/x',
          statusCode: 200,
          durationMs: 5,
          requestBodyBytes: 0,
          responseBodyBytes: 2,
          requestHeaders: {},
          responseHeaders: {},
          requestBody: null,
          responseBody: null,
          timestampMs: 0,
        );

    test('identify() emits an identify event and attaches external_id', () async {
      final sink = newSink();
      sink.identify('user_123', {'email': 'a@b.com', 'plan': 'pro'});
      await Future<void>.delayed(const Duration(milliseconds: 50));

      expect(bodies, isNotEmpty);
      final envelope = bodies.last;
      expect(envelope['external_id'], 'user_123');

      final events = (envelope['events'] as List).cast<Map<String, dynamic>>();
      final identifyEv = events.firstWhere((e) => e['kind'] == 'identify');
      expect(identifyEv['external_id'], 'user_123');
      expect(jsonEncode(identifyEv), isNot(contains('a@b.com')));
      expect(jsonEncode(identifyEv), contains('pro'));
    });

    test('empty externalId is a no-op (never throws)', () async {
      final sink = newSink();
      expect(() => sink.identify('', {'email': 'a@b.com'}), returnsNormally);
      await Future<void>.delayed(const Duration(milliseconds: 50));
      expect(bodies, isEmpty);
    });

    test('external_id attaches to a subsequent normal event', () async {
      final sink = newSink();
      sink.identify('user_456');
      await Future<void>.delayed(const Duration(milliseconds: 50));
      bodies.clear();

      sink.addNetwork(sampleEvent());
      await Future<void>.delayed(const Duration(milliseconds: 50));
      expect(bodies, isNotEmpty);
      expect(bodies.last['external_id'], 'user_456');
    });

    test('setExperiment attaches experiments; null clears; bad input is a no-op', () async {
      final sink = newSink();
      sink.setExperiment('checkout_flow', 'variant_b');
      sink.addNetwork(sampleEvent());
      await Future<void>.delayed(const Duration(milliseconds: 50));
      expect(bodies.last['experiments'], {'checkout_flow': 'variant_b'});

      expect(() => sink.setExperiment('', 'x'), returnsNormally); // bad input: no-op
      sink.setExperiment('checkout_flow', null);
      sink.addNetwork(sampleEvent());
      await Future<void>.delayed(const Duration(milliseconds: 50));
      expect(bodies.last.containsKey('experiments'), isFalse);
    });
  });
}
